install-usercss/install-usercss.js

```diff
@@ -391,47 +308,103 @@
     }
   }
 
-  function getCodeDirectly() {
-    // FF applies page CSP even to content scripts, https://bugzil.la/1267027
-    // To circumvent that, the bg process downloads the code directly
-    const key = 'tempUsercssCode' + tabId;
-    chrome.storage.local.get(key, data => {
-      const code = data && data[key];
-
-      // bg already downloaded the code
-      if (typeof code === 'string') {
-        initSourceCode(code);
-        chrome.storage.local.remove(key);
-        return;
-      }
-
-      // bg still downloads the code
-      if (code && code.loading) {
-        const waitForCodeInStorage = (changes, area) => {
-          if (area === 'local' && key in changes) {
-            initSourceCode(changes[key].newValue);
-            chrome.storage.onChanged.removeListener(waitForCodeInStorage);
-            chrome.storage.local.remove(key);
+  function initLiveReload() {
+    const DELAY = 500;
+    let isEnabled = false;
+    let timer = 0;
+    /** @type function():Promise<string|null> */
+    let getData = null;
+    /** @type Promise */
+    let sequence = null;
+    if (tabId < 0) {
+      getData = DirectDownloader();
+      sequence = API.getUsercssInstallCode(initialUrl).catch(getData);
+    } else {
+      getData = PortDownloader();
+      sequence = getData();
+    }
+    return {
+      get enabled() {
+        return isEnabled;
+      },
+      ready: sequence,
+      onToggled(e) {
+        if (e) isEnabled = e.target.checked;
+        if (installed || installedDup) {
+          (isEnabled ? start : stop)();
+          $('.install').disabled = isEnabled;
+          $('#live-reload-install-hint').hidden = !isEnabled;
+          $('#live-reload-install-hint-ff').hidden = !isEnabled || tabId < 0;
+        }
+      },
+    };
+    function check() {
+      getData(initialUrl)
+        .then(update, logError)
+        .then(() => {
+          timer = 0;
+          start();
+        });
+    }
+    function logError(error) {
+      console.warn(t('liveReloadError', error));
+    }
+    function start() {
+      timer = timer || setTimeout(check, DELAY);
+    }
+    function stop() {
+      clearTimeout(timer);
+      timer = 0;
+    }
+    function update(code) {
+      if (code == null) return;
+      sequence = sequence.then(() => {
```

Comment:
Once something throws `sequence` will be break. We should catch the error before adding a new promise:
```js
sequence = sequence.catch(console.error).then(() => {
  ...
})
```

Reply:
What exactly throws? The last part in sequence is `.catch(showError)` and if something throws there it would mean an unrecoverable bug anyway.

Reply:
Then let's move the `catch` to the outside to catch all errors:
```diff
diff --git a/install-usercss/install-usercss.js b/install-usercss/install-usercss.js
index 95916aee..569ea2d7 100644
--- a/install-usercss/install-usercss.js
+++ b/install-usercss/install-usercss.js
@@ -368,9 +368,8 @@
         cm.setCursor(cursor);
         cm.scrollTo(scrollInfo.left, scrollInfo.top);
         return API.installUsercss({id, sourceCode: code})
-          .then(updateMeta)
-          .catch(showError);
-      });
+          .then(updateMeta);
+      }).catch(showError);
     }
     function DirectDownloader() {
       let oldCode = null;

```

Reply:
showError catches for installUsercss, it doesn't catch for anything else. I'm getting tired of this so let's assume you were right initially.